core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java

```diff
@@ -195,15 +198,57 @@ public Result visit(Project e) {
     final Builder builder =
         x.builder(e, Clause.SELECT);
     final List<SqlNode> selectList = new ArrayList<>();
+
+    boolean checkNull = isNeedCastNull(stack);
     for (RexNode ref : e.getChildExps()) {
       SqlNode sqlExpr = builder.context.toSql(null, ref);
+      if (checkNull && SqlUtil.isNullLiteral(sqlExpr, false)) {
+        sqlExpr = castNullType(sqlExpr, e.getRowType().getFieldList().get(selectList.size()));
+      }
       addSelect(selectList, sqlExpr, e.getRowType());
     }
 
     builder.setSelect(new SqlNodeList(selectList, POS));
     return builder.result();
   }
 
+  /**
+   * Check whether null item in select need to be casted
+   * @param stack stack of visit
+   * @return If null need to be casted then return true, otherwise return false.
+   */
+  private static boolean isNeedCastNull(Deque<Frame> stack) {
+    final int stackSize = stack.size();
+    if (stackSize > 1) {
```

Comment:
If I am not mistaken, this `if` is not really needed, just having a for loop:
`for (int i = 1; i < stack.size(); i++) { ... }` should achieve the same result.
Also, we can get rid of `stackSize` local variable.

Reply:
1. Yes, "if" indeed need to be removed.
2. I think stackSize is needed because that there is extra computation in  java.util.ArrayDeque#size

Reply:
Yep, agree on keeping stackSize local variable